src/main/scala/cromwell/engine/io/gcs/NioGcsPath.scala

```diff
@@ -0,0 +1,104 @@
+package cromwell.engine.io.gcs
+
+import java.io.File
+import java.net.URI
+import java.nio.file.WatchEvent.{Kind, Modifier}
+import java.nio.file._
+import java.util
+
+import scala.collection.JavaConverters._
+import scala.language.implicitConversions
+import scala.language.postfixOps
+
+object NioGcsPath {
+  def apply(path: String)(implicit gcsFileSystem: GcsFileSystem)= {
+    gcsFileSystem.getPath(path)
+  }
+
+  implicit class PathEnhanced(val path: Path) extends AnyVal {
+    def asGcsPath(implicit gcsFileSystem: GcsFileSystem) = path match {
+      case gcsPath: NioGcsPath => gcsPath
+      case otherPath: Path => gcsFileSystem.getFlexiblePath(otherPath.toString).asInstanceOf[NioGcsPath]
+      case _ => throw new IllegalArgumentException("Only GcsPaths are supported.")
+    }
+  }
+}
+
+class NioGcsPath(private val chunks: Array[String], absolute: Boolean)(implicit gcsFileSystem: GcsFileSystem) extends Path {
+  import NioGcsPath._
+
+  private val SEPARATOR = gcsFileSystem.getSeparator
+  private val PROTOCOL = gcsFileSystem.PROTOCOL
```

Comment:
Scala conventions are to not have all caps, and as these aren't companion object vals I think they're supposed to be completely lowercase